File: 101-200/148-sort-list.cpp
ListNode* sortList(ListNode* head) {
    if (head == nullptr || head->next == nullptr) return head;
    ListNode *slow = head, *fast = head;
    ListNode *temp = nullptr;
    while (fast && fast->next) {
        temp = slow;
        slow = slow->next;
        fast = fast->next->next;
    }
    temp->next = nullptr;
    ListNode *left = sortList(head);
    ListNode *right = sortList(slow);

    ListNode *dummyHead = new ListNode(-1);
    dummyHead->next = head;
    ListNode *curr = dummyHead;
    while (left && right) {
        if (left->val <= right->val) {
            curr->next = left;
            left = left->next;
        }
        else {
            curr->next = right;
            right = right->next;
        }
        curr = curr->next;
    }

    if (left) curr->next = left;
    if (right) curr->next = right;
    return dummyHead->next;
}

File: python version/148-sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None: return head
        slow = fast = head
        prev = None
        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next.next
        tmp = prev.next
        prev.next = None
        sorted_left = self.sortList(head)
        sorted_right = self.sortList(tmp)
        dummy_head = ListNode(0)
        curr = dummy_head
        l = sorted_left
        r = sorted_right
        while l and r:
            if l.val < r.val:
                curr.next = l
                l = l.next
            else:
                curr.next = r
                r = r.next
            curr = curr.next
        if l:
            curr.next = l
        elif r:
            curr.next = r

        return dummy_head.next